Source/ParticleInterpolators/ParticleInterpolators.impl.hpp

```diff
@@ -0,0 +1,655 @@
+            if (variable_type == VariableType::state)
+            {
+                parity *= BoundaryConditions::get_state_var_parity(comp, dir);
+            }
+            else if (variable_type == VariableType::derived)
+            {
+                // if parity was not set for derived vars, print a message
+                AMREX_ALWAYS_ASSERT(m_derived_bc_parity[comp] !=
+                                    BCParity::undefined);
+                BCParity comp_parity  = m_derived_bc_parity[comp];
+                auto dir_parities     = bc_parity_map.at(comp_parity);
+                parity               *= dir_parities[dir];
+            }
+            // invert parity for first derivatives
+            if (deriv[dir] == 1)
+            {
+                parity *= -1;
+            }
+        }
+    }
+    return parity;
+}
+
+// a function to reflect a particle back into the valid domain, when symmetry
+// BCs are used
+template <int num_components>
+AMREX_GPU_HOST_DEVICE AMREX_FORCE_INLINE amrex::Real
+ParticleInterpolators<num_components>::reflect_particle(amrex::Real x,
+                                                        amrex::Real lo,
+                                                        amrex::Real hi,
+                                                        bool lo_reflect,
+                                                        bool hi_reflect)
+{
+    // enforce a new particle position if needed
+    amrex::Real xl = x;
+    if (lo_reflect && xl < lo)
+    {
+        xl = lo + (lo - xl); // reflect across lo
+    }
+    if (hi_reflect && xl > hi)
+    {
+        xl = hi - (xl - hi); // reflect across hi
+    }
+
+    return xl;
+}
+
+// allocate particles at the query points
+template <int num_components>
+void ParticleInterpolators<num_components>::populate_from_query(
+    const InterpolationQueryParticle &query)
+{
+    AMREX_ASSERT(m_initialized);
+
+    // we populate particles with rank 0
+    const bool amroot = (amrex::ParallelDescriptor::MyProc() == 0);
+
+    const int lev = 0;
+    const int n =
+        static_cast<int>(query.m_num_points); // number of query points
+
+    if (amroot)
+    {
+        // it does not matter on which level the particles are initialised so
+        // long as we Redistribute() them later
+        auto &ptile = this->DefineAndReturnParticleTile(lev, 0, 0);
+        ptile.resize(n);
+        auto particle_data = ptile.getParticleTileData();
+
+        // get coords from query
+        const double *x = query.m_coords[0];
+        const double *y = query.m_coords[1];
+#if AMREX_SPACEDIM == 3
+        const double *z = query.m_coords[2];
+#endif
+
+        // Run a check on coords you are interpolating on
+        for (int i = 0; i < n; i++)
+        {
+            std::array<double, AMREX_SPACEDIM> coords{x[i], y[i]
+#if AMREX_SPACEDIM == 3
+                                                      ,
+                                                      z[i]
+#endif
+            };
```

Comment:
This probably isn't clang-formatted correctly but you get the idea.
```suggestion
            std::array<double, AMREX_SPACEDIM>
                coords{AMREX_D_DECL(x[i], y[i], z[i])};
```